pc-relate/pcrelate.wdl

```diff
@@ -0,0 +1,395 @@
+		# but it also is a not-optional input elsewhere -- double check!
+		File block = glob("*.RData")[0]
+	}
+}
+
+task pcrelate_correct {
+	input {
+		Array[File] pcrelate_block_files
+		Int n_sample_blocks = 1
+		Float sparse_threshold = 0.02209709
+
+		# runtime attributes
+		Int addldisk = 1
+		Int cpu = 2
+		Int memory = 4
+		Int preempt = 3
+	}
+	# Estimate disk size required
+	Int block_size = ceil(size(pcrelate_block_files, "GB"))
+	Int finalDiskSize = 2*block_size + addldisk
+
+	command <<<
+		set -eux -o pipefail
+
+		BASH_FILES=(~{sep=" " pcrelate_block_files})
+		for BASH_FILE in ${BASH_FILES[@]};
+		do
+			ln -s ${BASH_FILE} .
+		done
+
+		python << CODE
+		import os
+		f = open("pcrelate_correct.config", "a")
+		one_file = ["~{sep="," pcrelate_block_files}"][0]
+		prefix = os.path.basename(one_file).split("_block_")[0]
+		f.write("pcrelate_prefix \"%s\"\n" % prefix)
+		f.write("n_sample_blocks ~{n_sample_blocks}\n")
+		f.write("sparse_threshold ~{sparse_threshold}\n")
+		CODE
+
+		R -q --vanilla --args pcrelate_correct.config < /usr/local/analysis_pipeline/R/pcrelate_correct.R
+
+	>>>
+	
+	runtime {
+		cpu: cpu
+		docker: "uwgac/topmed-master@sha256:0bb7f98d6b9182d4e4a6b82c98c04a244d766707875ddfd8a48005a9f5c5481e"
+		disks: "local-disk " + finalDiskSize + " HDD"
+		memory: "${memory} GB"
+		preemptibles: "${preempt}"
+	}
+	output {
+		# CWL seems to have inconsistency where this is marked an optional output
+		# but it also is a not-optional input into kinship_plots -- double check!
+		File pcrelate_output = glob("*_pcrelate.RData")[0]
+		File pcrelate_matrix = glob("*_pcrelate_Matrix.RData")[0]
+	}
+}
+
+task kinship_plots {
+	input {
+		File kinship_file
+		String? kinship_method = "pcrelate"  # hardcoded on CWL?
+		Float? kinship_plot_threshold
+		File? phenotype_file
+		String? group
+		File? sample_include_file
+		String out_prefix_initial = ""
+		
+		# runtime attributes
+		Int addldisk = 1
+		Int cpu = 2
+		Int memory = 4
+		Int preempt = 3
+	}
+	String out_prefix_final = out_prefix_initial + "_pcrelate"
+	
+	command {
+		set -eux -o pipefail
+
+		python << CODE
+		import os
+		f = open("kinship_plots.config", "a")
+		f.write('kinship_file "~{kinship_file}"\n')
+		if "~{kinship_method}" not in ['pcrelate', 'king_ibdseg', 'king_robust']:
+			f.close()
+			print("Invalid kinship method. Must be pcrelate, king_robust, or king_ibdseg")
+			exit(1)
+		elif "~{kinship_method}" == "king_robust":
+			f.write('kinship_method "king"\n')
+		else:
+			f.write('kinship_method "~{kinship_method}"\n')
+
+		if "~{kinship_plot_threshold}" is not "":
+			f.write('kinship_threshold "~{kinship_plot_threshold}"\n')
+
+		f.write('out_file_all "~{out_prefix_final}_all.pdf"\n')
+		f.write('out_file_cross "~{out_prefix_final}_cross_group.pdf"\n')
+		f.write('out_file_study "~{out_prefix_final}_within_group.pdf"\n')
+
+		if "~{phenotype_file}" is not "":
+			f.write('phenotype_file "~{phenotype_file}"\n')
```

Comment:
In the past, you have used a call to defined() and then the bool value defPhenotypeFile. Is there a reason we switched?

Reply:
I switched for two reasons:
1. I don't recall the details, but the defined() check did bug out on me once on Terra. This might be related to [the basename() bug involving DRS](https://support.terra.bio/hc/en-us/community/posts/4405396480027-basename-does-not-function-as-expected-on-DRS-URIs)?
2. Doing it this way avoids having to create one new boolean per File? input and is therefore a bit cleaner.